doc/stm/index.mld

```diff
@@ -0,0 +1,583 @@
+    | Cardinal, Res ((Int,_), l)              -> l = List.length state
+    | Add _,  Res ((Unit,_),_)                -> true
+    | Remove i, Res ((Option Int, _), Some x) -> List.mem i state && i = x
+    | Remove i, Res ((Option Int, _), None)   -> not (List.mem i state)
+    | _                                       -> false
+]}
+
+Here, we can pattern match on the result of the [Remove i] [cmd] and express
+the fact that if this result is [Some x] then [i] was indeed in the set and [i]
+and [x] are structurally equal. On the other hand, if the result is [None], then
+[i] was not in the set.
+
+Now, let's say we have updated [arb_cmd] just by adding a generator for the [Remove]
+[cmd] similar to the other ones:
+
+{[
+  let arb_cmd _state =
+    QCheck.make ~print:show_cmd
+      (QCheck.Gen.oneof
+        [Gen.return Cardinal;
+         Gen.map (fun i -> Mem i) Gen.int;
+         Gen.map (fun i -> Add i) Gen.int;
+         Gen.map (fun i -> Remove i) Gen.int;
+        ])
+]}
+
+Then, if we make the same mistake as above for the [add] function, namely:
+
+{[
+  let remove a t =
+    Mutex.lock t.mutex;
+    let r =
+      if mem_non_lock a t then begin
+        t.content <- S.remove a t.content;
+        (* t.cardinal <- t.cardinal - 1; *)
+        Some a
+      end
+      else None
+    in
+    Mutex.unlock t.mutex;
+    r
+]}
+
+most of the time, [qcheck-stm] won't be able to spot the bug:
+
+{[
+$ dune exec ./stm-doc/mutable_set_v4.exe
+random seed: 423411827
+generated error fail pass / total     time test name
+[✓]  100    0    0  100 /  100     0.8s STM sequential tests
+================================================================================
+success (ran 1 tests)
+]}
+
+The reason why is that there is very little chance to generate an argument for
+[Remove] corresponding to an element already added to the set in the same
+program. So the conditional branch where the mistake has been made is never
+explored. But be sure that it will be in real life!
+
+In order to overcome this problem, it is possible to skew the distribution of
+generated [cmd]s by looking at the [state] parameter. If the model of the set
+is empty, then we just generate a random element. But if there are already some
+elements in the set, then we can choose between picking one of them or
+generating a random new one. Now that we are aware of the problem, we also
+update the generators for [Mem] and [Add] which suffer from the same weakness.
+
+{[
+  let arb_cmd state =
+    let gen =
+      match state with
+      | [] -> Gen.int
+      | xs -> Gen.(oneof [oneofl xs; int])
+    in
+    QCheck.make ~print:show_cmd
+      (QCheck.Gen.oneof
+        [Gen.return Cardinal;
+         Gen.map (fun i -> Mem i) gen;
+         Gen.map (fun i -> Add i) gen;
+         Gen.map (fun i -> Remove i) gen;
+        ])
+]}
+
+This will be enough to trigger a failure in the test which can help us correct
+our implementation.
+
+{[
+random seed: 284906494
```

Comment:
The recent commit c598032 introduced a duplicate seed.
This one can be removed.